[AMDGPU][MC] Implement fft and rotate modes for ds_swizzle_b32

jwanggit86 opened the pull request:
In addition to the basic mode, the ds_swizzle_b32 is supposed to support two specific modes: fft and rotate. This patch implements those two modes.

jwanggit86 committed:
[AMDGPU][MC] Implement fft and rotate modes for ds_swizzle_b32

In addition to the basic mode, the ds_swizzle_b32 is supposed to
support two specific modes: fft and rotate. This patch implements
those two modes.

DadSchoorse commented:
I think rotate/fft are unsupported on GFX6/7, so using them should probably result in assembler errors for those generations. (and maybe GFX8, though someone in #28975 said it's supported there, just not mentioned in the documentation)

DadSchoorse commented on llvm/lib/Target/AMDGPU/AsmParser/AMDGPUAsmParser.cpp:
```diff
@@ -8141,6 +8143,56 @@ AMDGPUAsmParser::parseSwizzleBitmaskPerm(int64_t &Imm) {
+  if (!parseExpr(Type))
+    return false;
+
+  if (Type != FFT_NO_SWIZZLE && Type != FFT_SWIZZLE_00 &&
+      Type != FFT_SWIZZLE_10) {
+    const std::string ErrMsg = "invalid FFT swizzle type: must be " +
+                               std::to_string(FFT_SWIZZLE_00) + ", " +
+                               std::to_string(FFT_SWIZZLE_10) + ", or " +
+                               std::to_string(FFT_NO_SWIZZLE);
```
Are you sure that only these three modes are supported? The pseudo code in the ISA docs would work for any value in [0:31].

jwanggit86 replied:
Good point. It's possible the ISA doc lists 0x00, 0x10, and 0x1f as examples, not enumeration of allowed values. Will fix.

arsenm commented on llvm/lib/Target/AMDGPU/AsmParser/AMDGPUAsmParser.cpp:
```diff
@@ -8141,6 +8143,56 @@ AMDGPUAsmParser::parseSwizzleBitmaskPerm(int64_t &Imm) {
+    return false;
+
+  int64_t RotateSize;
+  const std::string ErrorMsg =
+      "number of threads to rotate must be in the interval [0," +
+      std::to_string(ROTATE_MAX_SIZE) + "]";
```
shouldn't need to pre-create the error string, also these all use Twine?

jwanggit86 committed:
Fix a mistake in previous commit: for FFT mode, the swizzle value
is in [0,31].

jwanggit86 committed:
This commit: (1) disallow the words FFT/ROTATE in ds_swizzle_b32
for pre-gfx9 targets (2) modify code to align with ISA doc, namely,
if offset >= 0xE000 then it is the FFT mode, and so on.

jwanggit86 commented:
> I think rotate/fft are unsupported on GFX6/7, so using them should probably result in assembler errors for those generations. (and maybe GFX8, though someone in #28975 said it's supported there, just not mentioned in the documentation)

FFT/ROTATE are now disallowed for pre-gfx9 targets.

arsenm commented on llvm/lib/Target/AMDGPU/MCTargetDesc/AMDGPUInstPrinter.cpp:
```diff
@@ -1503,8 +1503,21 @@ void AMDGPUInstPrinter::printSwizzle(const MCInst *MI, unsigned OpNo,
+  // Rotate and FFT modes
+  if (Imm >= ROTATE_MODE_LO && AMDGPU::isGFX9Plus(STI)) {
+    if (Imm >= FFT_MODE_LO) {
+      O << "swizzle(" << IdSymbolic[ID_FFT] << "," << (Imm & FFT_SWIZZLE_MASK)
+        << ")";
+    } else if (Imm >= ROTATE_MODE_LO) {
+      O << "swizzle(" << IdSymbolic[ID_ROTATE] << ","
+        << ((Imm >> ROTATE_DIR_SHIFT) & ROTATE_DIR_MASK) << ","
+        << ((Imm >> ROTATE_SIZE_SHIFT) & ROTATE_SIZE_MASK) << ")";
```
```suggestion
      O << "swizzle(" << IdSymbolic[ID_FFT] << ',' << (Imm & FFT_SWIZZLE_MASK)
        << ')';
    } else if (Imm >= ROTATE_MODE_LO) {
      O << "swizzle(" << IdSymbolic[ID_ROTATE] << ','
        << ((Imm >> ROTATE_DIR_SHIFT) & ROTATE_DIR_MASK) << ','
        << ((Imm >> ROTATE_SIZE_SHIFT) & ROTATE_SIZE_MASK) << ')';
```

jwanggit86 replied:
Done.

arsenm reviewed (commented):
lgtm with nits

jwanggit86 committed:
This commit: (1) for parseSwizzleOperand(), change the type of
ErrMsg from StringRef to Twine (2) change some single-char string
to single quotes (3) undo some changes by clang-format.

jwanggit86 commented on llvm/lib/Target/AMDGPU/Utils/AMDGPUAsmUtils.cpp:
```diff
@@ -646,12 +646,9 @@ unsigned const DfmtNfmt2UFmtGFX11[] = {
-  "REVERSE",
-  "BROADCAST",
+const char *const IdSymbolic[] = {
+    "QUAD_PERM", "BITMASK_PERM", "SWAP",   "REVERSE",
```
Done.

jwanggit86 commented on llvm/lib/Target/AMDGPU/AsmParser/AMDGPUAsmParser.cpp:
```diff
@@ -8141,6 +8143,54 @@ AMDGPUAsmParser::parseSwizzleBitmaskPerm(int64_t &Imm) {
+  SMLoc Loc;
+  if (!parseSwizzleOperand(Swizzle, 0, FFT_SWIZZLE_MAX,
+                           "FFT swizzle must be in the interval [0," +
+                               Twine(FFT_SWIZZLE_MAX).str() + Twine(']').str(),
```
Done.

jwanggit86 commented on llvm/lib/Target/AMDGPU/AsmParser/AMDGPUAsmParser.cpp:
```diff
@@ -8141,6 +8143,54 @@ AMDGPUAsmParser::parseSwizzleBitmaskPerm(int64_t &Imm) {
+  if (!parseSwizzleOperand(
+          RotateSize, 0, ROTATE_MAX_SIZE,
+          "number of threads to rotate must be in the interval [0," +
+              Twine(ROTATE_MAX_SIZE).str() + Twine(']').str(),
```
Done.

arsenm commented on llvm/lib/Target/AMDGPU/AsmParser/AMDGPUAsmParser.cpp:
```diff
@@ -1829,7 +1829,7 @@ class AMDGPUAsmParser : public MCTargetAsmParser {
                            const unsigned MinVal,
                            const unsigned MaxVal,
-                           const StringRef ErrMsg,
+                           const Twine ErrMsg,
```
const Twine &

arsenm commented on llvm/lib/Target/AMDGPU/AsmParser/AMDGPUAsmParser.cpp:
```diff
@@ -7977,7 +7977,7 @@ bool
                                      const unsigned MinVal,
                                      const unsigned MaxVal,
-                                     const StringRef ErrMsg,
+                                     const Twine ErrMsg,
```
const Twine &

jwanggit86 replied:
done.

arsenm reviewed (commented):
lgtm with const references to twine

jwanggit86 committed:
For function parseSwizzleOperand(), change type of ErrMsg from
const Twine to const Twine&. Also add "clang-format off" to
prevent clang-format make unwanted format changes.

jayfoad commented on llvm/test/MC/AMDGPU/ds_swizzle.s:
```diff
@@ -0,0 +1,131 @@
+// RUN: not llvm-mc -triple=amdgcn -mcpu=bonaire -show-encoding %s | FileCheck -check-prefix=GFX7 %s
+// RUN: not llvm-mc -triple=amdgcn -mcpu=tonga -show-encoding %s | FileCheck -check-prefix=GFX8 %s
+// RUN: not llvm-mc -triple=amdgcn -mcpu=gfx908 -show-encoding %s | FileCheck %s
```
```suggestion
// RUN: not llvm-mc -triple=amdgcn -mcpu=gfx908 -show-encoding %s | FileCheck -check-prefix=GFX9 %s
```
Please also add some tests for disassembly of the ROTATE form.

jwanggit86 replied:
Both done.

jwanggit86 committed:
Update tests.

DadSchoorse commented on llvm/lib/Target/AMDGPU/AsmParser/AMDGPUAsmParser.cpp:
```diff
@@ -8141,6 +8138,54 @@ AMDGPUAsmParser::parseSwizzleBitmaskPerm(int64_t &Imm) {
+bool AMDGPUAsmParser::parseSwizzleRotate(int64_t &Imm) {
+  using namespace llvm::AMDGPU::Swizzle;
+
+  if (!AMDGPU::isGFX9Plus(getSTI())) {
```
I now tested that these patterns do indeed work on gfx8, and radv will start using them there. So please consider changing this to gfx8+.

jwanggit86 replied:
Based on the ISA doc, these two modes are not supported on GFX8.

DadSchoorse replied:
Given the information in #28975, that looks to be a mistake in the public ISA doc. Also, I confirmed that these instructions pass the subgroupRotate Vulkan-CTS tests on polaris10.

jwanggit86 replied:
> Given the information in #28975, that looks to be a mistake in the public ISA doc. Also, I confirmed that these instructions pass the subgroupRotate Vulkan-CTS tests on polaris10.

@arsenm Your thoughts?

jayfoad replied:
@DadSchoorse I believe this is one of those cases that might happen to work, but for whatever reason was never fully verified in GFX8 hardware. So you can use it at your own risk if you want, but the decision not to document it was deliberate, and for the same reason I would be reluctant to support it in the LLVM assembler/disassembler.

Venemo replied:
FWIW, they seem to work on Fiji also.

> was never fully verified in GFX8 hardware

Can you please elaborate what that means exactly?
